util/src/main/java/com/google/cloud/hadoop/util/ThreadTrace.java

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright 2023 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.hadoop.util;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.gson.Gson;
+import com.google.gson.annotations.Expose;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Helps track a request. For eg. this can be used to track a HCFS API call (e.g. rename). This can
+ * help pinpoint the operation (e.g. a GCS API call) which is causing request to be slow.
+ *
+ * <p>This also tracks operations which runs in the same thread or another thread, up to one level.
+ * Mmeaning if an main thread uses threadA, that is tracked. But if threadA uses threadB, it is not
+ * tracked. This case can be added if required.
+ *
+ * <p>NOTE: Do not directly use this class. Use the factory methods in {@code ThreadOperation}. This
+ * class is for internal use and the API signatures can change.
+ */
+public class ThreadTrace {
+
+  // TODO: Make this configurable
+  static final int MAX_EVENTS_SIZE = 20;
+  private static Gson gson = new Gson();
+  private final String trackingId;
+  private final List<ThreadTraceEvent> events;
+
+  private final Map<Long, List<ThreadTraceEvent>> subEvents = new HashMap<>();
+
+  public ThreadTrace(String trackingId) {
+    this(trackingId, new ArrayList<>());
+  }
+
+  private ThreadTrace(String trackingId, List<ThreadTraceEvent> events) {
+    checkNotNull(trackingId, "trackingId cannot be null");
+    checkNotNull(events, "events cannot be null. trackingId=%s", trackingId);
+    this.trackingId = trackingId;
+    this.events = events;
+  }
+
+  @Override
+  public String toString() {
+    return gson.toJson(this);
+  }
+
+  ThreadTraceEvent starOperationEvent(String name) {
+    if (isFull()) {
+      return null;
+    }
+
+    ThreadTraceEvent theEvent = ThreadTraceEvent.startEvent(name);
+    this.events.add(theEvent);
+
+    return theEvent;
+  }
+
+  void endOperationEvent(ThreadTraceEvent startEvent) {
+    if (isFull() || startEvent == null) {
+      return;
+    }
+
+    int eventsSize = events.size();
+    if (eventsSize != 0) {
+      ThreadTraceEvent lastEvent = events.get(eventsSize - 1);
```

Comment:
One doubt: why we need to match it with the most recent event(or last) in array? We have reference to the startEvent and now an End is been invoked on it. Why not just go ahead and update the type of `START` event to be of type `MERGED` instead off creating an END type event?

Reply:
For eg. op1.start -> op2.start -> op2.end ->op1.end It is good to know the tree structure of the operations.